torchok/tasks/detection.py

```diff
@@ -0,0 +1,106 @@
+from typing import Any, Dict, Union
+
+import torch
+import torch.nn as nn
+from omegaconf import DictConfig
+
+from torchok.constructor import BACKBONES, DETECTION_NECKS, HEADS, TASKS
+from torchok.constructor.config_structure import Phase
+from torchok.models.backbones import BackboneWrapper
+from torchok.tasks.base import BaseTask
+
+
+@TASKS.register_class
+class SingleStageDetectionTask(BaseTask):
+    def __init__(self, hparams: DictConfig):
+        """Init SingleStageDetectionTask.
+
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__(hparams)
+
+        # BACKBONE
+        backbone_name = self._hparams.task.params.get('backbone_name')
+        backbones_params = self._hparams.task.params.get('backbone_params', dict())
+        self.backbone = BACKBONES.get(backbone_name)(**backbones_params)
+        self.num_scales = self._hparams.task.params.get('num_scales', len(self.backbone.out_encoder_channels))
+
+        # NECK
+        neck_name = self._hparams.task.params.get('neck_name')
+        neck_params = self._hparams.task.params.get('neck_params', dict())
+        neck_in_channels = self.backbone.out_encoder_channels[-self.num_scales:][::-1]
+        self.neck = DETECTION_NECKS.get(neck_name)(in_channels=neck_in_channels, **neck_params)
+
+        # HEAD
+        head_name = self._hparams.task.params.get('head_name')
+        head_params = self._hparams.task.params.get('head_params', dict())
+        head_in_channels = self.neck.out_channels
+        self.bbox_head = HEADS.get(head_name)(in_channels=head_in_channels, **head_params)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward method."""
+        x = self.backbone.forward_features(x)[-self.num_scales:]
+        with torch.cuda.amp.autocast(dtype=torch.float32):
+            x = self.neck(x)
+        x = self.bbox_head(x)
+        return x
+
+    def forward_with_gt(self, batch: Dict[str, Union[torch.Tensor, int]]) -> Dict[str, Any]:
+        """Forward with ground truth labels."""
+        input_data = batch.get('image')
+        features = self.backbone.forward_features(input_data)[-self.num_scales:]
```

Comment:
Our forward feature reurn [x] + features, is mmdet also need x?

Reply:
No, here we take the features from the latest backbone stages like `[layer2, layer3, layer4]`